### DISCUSSION TITLE
[Question][Hw5] Several questions about abstract semantics

### DISCUSSION THREAD

**Discussion by SangHyeon-Lee** [prosyslab-classroom/cs348-information-security#622] [Q&A]
Hello, I have some questions about abstract semantics in hw5.

1. In the instruction, there is assignment node in atomic node. Is there corresponding Opcode for assignment? Furthermore, for example, in the instruction x = 2, how can I get x and 2 in transfer_atomic function?
2.  What is the difference between transfer_cond and filter(in semantic.ml) function?
3. It seems that we can get PHI nodes from for loop in example4.c, what is the reason of getting PHI nodes from for loop? In what specific situations PHI nodes invoke?

Thank you in advance!

**Comment by RiceBiscuits** [comment]
1. LLVM IR has some instruction about assignments, but you don't need to consider the instruction in this homework. 
Because LLVM usually optimize c code and an integer assignment also will be disappeared.
Please look at any example IR file. IR will put an immediate value into instructions (e.g., add)directly instead of reference a variable that store an integer.
2. Please check this issue #166

**Comment by RiceBiscuits** [comment]
3. To choose valid version of a variable that modified at the branched basic blocks.
Please refer this page (https://en.wikipedia.org/wiki/Static_single_assignment_form)
for and if-else condition can invoke PHI nodes.

**Comment by SangHyeon-Lee** [comment]
Thank you for your answers!
I'm still curious about question 1.
For example, of add node, x = y + z, I think I should add (x -> abstract value of addition) in memory. Without referencing variable, how can I update the memory?

**Comment by RiceBiscuits** [comment]
> Thank you for your answers!
> I'm still curious about question 1.
> For example, of add node, x = y + z, I think I should add (x -> abstract value of addition) in memory. Without referencing variable, how can I update the memory?

Your analyzer will read an IR file instead of c source code. 
That means, source code `x = y + z` , which is written in c grammar, is no longer exist.

```
y = 1;
z = 2;
x = y + z;
```

This code will be translated into:
`%x = add 1 2 `
like this. Variable y and z are disappeared, only immediate value 1, 2 are given.  
You can read the immediate value instead of the variables, and use it for updating a memory.

**Comment by SangHyeon-Lee** [comment]
Sorry for keep asking....
Then how can I update x? How can I know I should update "x" from that instruction?

Thank you very much!!

**Comment by RiceBiscuits** [comment]
Do you mean which instruction should we deal?
or
How to change an immediate value to Value.t?
or
Other thing?

**Comment by SangHyeon-Lee** [comment]
So we need to add x -> 3 into memory, 
from val add : Variable.t -> Value.t -> t -> t, we need to specify variable name, which is x in above example.
My question was how can I get variable "x" from instruction `%x = add 1 2`

Thank you

**Comment by RiceBiscuits** [comment]
Maybe you already have Variable x.
It's argument `instr` in transfer_* functions.

**Comment by KihongHeo** [comment]
See [this doc](https://github.com/prosyslab-classroom/llvm-primer/blob/master/data-structure.md).

**Comment by SangHyeon-Lee** [comment]
Oh, I got it!
Thank you very much
